### ISSUE TITLE
[$250] [HOLD for payment 2024-11-14] [Search v2.4] - Same search query appears twice in Recent searches

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#51044]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.50-3
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** applausetester+kh081006@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
Precondition:
- User has not searched using filters previously
1. Go to staging.new.expensify.com
2. Go to Search
3. Click Filters
4. Apply any filter and click View results
5. Go to Inbox
6. Click on the search icon on the report header
7. Click on the search query under Recent searches
8. Go back to Inbox
9. Click on the search icon on the report header

## Expected Result:
The search query made in Step 4 will only appear once after clicking on the same search query again in Step 7

## Actual Result:
The search query made in Step 4 appears twice after clicking on the same search query again in Step 7

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/4721e008-4faa-4216-ab1f-2561b9dd4a03

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021857076660400822469</li>
        <li>Upwork Job ID: 1857076660400822469</li>
        <li>Last Price Increase: 2024-11-14</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave-control

**Comment by luacmartins** [comment]
Interesting, the Onyx data is different from what's being displayed
![Screenshot 2024-10-17 at 11 22 32 AM](https://github.com/user-attachments/assets/04a7dd00-63b7-4bdb-bae5-78144cad27aa)

**Comment by jaydamani** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Same search query appears twice in Recent searches

### What is the root cause of that problem?

When we search again, new entry is added to nvp_RecentSearches

### What changes do you think we should make in order to solve the problem?

Add logic to remove duplicate queries and keep only most recent one in below line.

https://github.com/Expensify/App/blob/2aae716653057688ea752e0d47903c11db9d03fc/src/components/Search/SearchRouter/SearchRouter.tsx#L55

**Comment by jaydamani** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Same search query appears twice in Recent searches

### What is the root cause of that problem?

When we search again, new entry is added to nvp_RecentSearches

### What changes do you think we should make in order to solve the problem?

Add logic to remove duplicate queries and keep only most recent one in below line.

https://github.com/Expensify/App/blob/2aae716653057688ea752e0d47903c11db9d03fc/src/components/Search/SearchRouter/SearchRouter.tsx#L55

**Comment by dangrous** [comment]
@luacmartins do you want to take this one since you know the search stuff well?

**Comment by SzymczakJ** [comment]
I'd like to work on it, since I was implementing Search 2.4.

**Comment by luacmartins** [comment]
@SzymczakJ all yours. I think we can demote this to NAB since it doesn't really break any functionality.

**Comment by SzymczakJ** [comment]
This bug is on FE
> Interesting, the Onyx data is different from what's being displayed

but after investigating this I think that it should be also fixed on BE. Recent Searches are generated on BE after searching for query. In case when we are searching for the same query twice, then it should be renewed in history and not duplicated. WDYT @luacmartins

**Comment by luacmartins** [comment]
@SzymczakJ that is the current logic in the backend. As you can see the Onyx data is accurate, not sure why the Onyx data and the list are out of sync
![377569515-04a7dd00-63b7-4bdb-bae5-78144cad27aa](https://github.com/user-attachments/assets/ef926c35-52f9-4997-be54-e05311acec7c)

**Comment by luacmartins** [comment]
Oh I see that it shows the query because we have instances of the same query saved with and without `sortBy:date sortOrder:desc`. We must be adding that in the frontend when sending it back to the server so we end up saving it twice instead of updating the query.

**Comment by luacmartins** [comment]
Hmm in the case of different sortBy or sortOrder, wouldn't we want to display that to the user, which would make it a different recent search

**Comment by luacmartins** [comment]
That sounds good to me 👍

**Comment by SzymczakJ** [comment]
After talking about this with @Kicu and @289Adam289 I'd like to propose another solution:
**Fix BE so same queries with changed sortOrder or sortBy should be treated the same way.** Here are arguments for that solution:
1. In development changing searchOrder is technically a new search, but for user it doesn't make much difference, take a look at this video as an example:

https://github.com/user-attachments/assets/e1953e01-3d42-4f48-951e-34429babc765

Now after playing around with sortOrder and sortBy for a couple of times user gets a list of the same query just with different sortOrder etc.. His meaningful previous searches are pushed out of recent list by garbage which is only differing in searchOrder. 🥲 

2. Having SearchOrder in recent searches list can lead some UI/UX problems because Search router has limited space horizontally and we won't have enough space to view even medium length queries.
<img width="540" alt="Screenshot 2024-10-21 at 13 01 00" src="https://github.com/user-attachments/assets/9388b0aa-74c8-42f0-b8ae-db74f8c86aeb">

This query has only one filter currency:PLN and even this one  filter isn't visible 😄
3. The same goes for Saved searches:
<img width="364" alt="Screenshot 2024-10-21 at 13 07 09" src="https://github.com/user-attachments/assets/192baa0b-2464-4deb-adc4-ae5cf7dfdc7b">

4. And everything that is shown on mobile screen:
<img width="375" alt="Screenshot 2024-10-21 at 13 08 18" src="https://github.com/user-attachments/assets/d5ae02e9-a45b-4583-aabe-2a24f5f07652">

After fixing this on BE we could think of some follow-ups like:
- change mobile design a little, so it allows user to sort search results on mobile screen(add some button for this) without writing it manually into SearchRouter. Now it's not possible.
- change designs a little bit, so recent searches and saved searches, can be displayed reasonably, because now only really short queries can be shown fully and showing any longer query is confusing.

**Comment by luacmartins** [comment]
I'm not sure that this is a backend fix. Currently, the logic in the backend is:

1. If canned search, return
2. Get `currentRecentSearches`
3. If `currentRecentSearches` has the request `hash`, update its timestamp
4. Else if `currentRecentSearches` > 5, purge oldest and add new search keyed by request hash
5. Else add the search to `currentRecentSearches` keyed by request hash

So the backend logic is already updating the timestamp. The issue is that we use the hash as the key and that's changing with sortBy/order. Is there any benefit in changing the hash when sortBy/sortOrder changes? We exclude `offset` from the hash already, so maybe we should be doing the same for sorting

**Comment by Kicu** [comment]
> Is there any benefit in changing the hash when sortBy/sortOrder changes?

A few things we'd need to check.
1. main thing is this - does the `hash` take any part in caching some parts of query or checking if the query is unique? I have completely misremembered that the hash is used as key in onyx, but after checking that does not seem to be the case
2. I'm pretty sure hash takes some part in deciding uniqueness of saved search, and highlighting if saved search is active
3. it's used as key in onyx for `optimisticData` for action `saveSearch` and `deleteSavedSearch` (something related to RBR which I don't fully understand) here: https://github.com/Expensify/App/blob/main/src/libs/actions/Search.ts#L63

So while this is one possible course of action, we'd need to verify if above cases don't break. Especially about point 1 - please remind me, we have never used hash as onyx key in front right? I imagined that? 😅

---
Now about what @SzymczakJ was trying to convey I think is this:
In case we keep hash as is right now, backend can see that the hash is different - so query is different.
However backend could *additionally* check if the different part of query is only `sort` values, and in that case don't treat it as new query; don't purge; just update the timestamp.

So basically for the user let's not save the same exact query that differs only by sorting.
However technically for us the query IS different, that's why I'm kinda reluctant to change `hash` generation.

In general we would like to avoid `Recent Searches` to look like this:
```
 | "sameQuery sortBy:date asc"
 | "sameQuery sortBy:date desc"
 | "sameQuery sortBy:from asc"
 | "sameQuery sortBy:from desc"
 | "sameQuery sortBy:date asc"
```
IMO ☝️ looks not really helpful. I'd rather see my actual older searches than the same search with changed sorts.

**Comment by SzymczakJ** [comment]
I think we have to include sortOrder and sortBy in hash because:
1. it is used as Onyx key on every search action. In previous hash function we also included it in hash because when the text is parsed the default sortOrder and sortBy are added. 
2. changing sortOrder and sortBy can give us two different results because of pagination(here I explained more https://github.com/Expensify/App/pull/50921#discussion_r1806361272) 

Also, fixing this on BE would be a problem, so I think the only solution we have is to have two hashes, one for computing recentSearches(which doesn't include sortOrder and sortBy) and the other one for everything else(this hash would include sort). @rayane-djouah proposed it in https://github.com/Expensify/App/issues/51044#issuecomment-2422773064.
WDYT @luacmartins @rayane-djouah

**Comment by luacmartins** [comment]
Yea, I think the simplest solution would be to have two hashes. We could add `recentSearchHash` to the top level of the jsonQuery object and in the backend we use `recentSearchHash` instead of `hash` to key recent searches. Otherwise the alternative solutions would be:

1. Remove sort params for hash. This would lead to us losing caching based on sorting and maybe lead to other unknown issues now.
2. Having to store the full jsonQuery in the database (instead of the query input string) and doing a deep object comparison with each search.

I like just adding the `recentSearchHash` and using it. Let's just leave a comment explaining what that hash does.

**Comment by Kicu** [comment]
Agreed that in the current situation storing two hashes seems the most reasonable 👍 

Feels to me like:
 - one hash is more "data oriented" - because sorting produces different datasets, we need different keys for them
 - the second hash will be more "user oriented" - so that user doesn't see (what looks like) identical queries

**Comment by johncschuster** [comment]
Payment Summary:

Contributor: @SzymczakJ does not require payment
Contributor+: @rayane-djouah paid $250 via Upwork

Upwork job [here](https://www.upwork.com/jobs/~021857076660400822469)! Please apply

**Comment by johncschuster** [comment]
Payment has been issued! Thanks for your contributions!
